### PR TITLE
Add basic line chart

### PR DESCRIPTION
**Pull request by naturetime10**
image

### CODE REVIEW

File: src/libs/chart/visual.ts [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            relativeX <= xRange.end;
+            relativeX += gridXGap
+        ) {
+            const line = new Line({
+                left: coordTranslator.xCoordToDrawRegion(relativeX),
+                top: coordTranslator.yCoordToDrawRegion(yRange.start),
+            });
+            line.strokeColor = color;
+            line.width = coordTranslator.yDistanceToDots(yRangeLen);
+            line.rotationAngle = 90;
```
Comment (duanxianbo) [inline src/libs/chart/visual.ts:59]: can we pass those to the Line constructor?

File: src/libs/chart/visual.ts [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    yLabelTitle: string,
+    yLabels: string[],
+    xLabelGap: number,
+    yLabelGap: number,
+): Drawable {
+    const drawables: Drawable[] = [];
+    const labelTextMeter = new TextMeter(ctx, '16px bold serif');
+    for (let index = 0; index < yLabels.length; index++) {
+        const yLabel = yLabels[index];
+
```
Comment (duanxianbo) [inline src/libs/chart/visual.ts]:
```suggestion
```

File: src/libs/chart/visual.ts [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    drawables.push(xLabelTitleText);
+    return new Group(drawables);
+}
+
+export function makeDataPointMarkers(
+    coordTranslator: CoordTranslator,
+    dataPoints: DataPoint[],
+): Drawable {
+    const drawables: Drawable[] = [];
+    for (let dataPoint of dataPoints) {
```
Comment (duanxianbo) [inline src/libs/chart/visual.ts]:
```suggestion
    for (const dataPoint of dataPoints) {
```

File: src/libs/chart/visual.ts [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+            left:
+                coordTranslator.xCoordToDrawRegion(dataPoint.x) -
+                markerSize.width / 2,
+            top:
+                coordTranslator.yCoordToDrawRegion(dataPoint.y) -
+                markerSize.height / 2,
+        });
+        marker.size = markerSize;
+        marker.strokeColor = undefined;
+        marker.fillColor = '#1976d2';
```
Comment (duanxianbo) [inline src/libs/chart/visual.ts:287]: can we put those colors into the visual file?

Comment (naturetime10) [reply src/libs/chart/visual.ts:287]: Do you meant to pass in as option?

Comment (duanxianbo) [reply src/libs/chart/visual.ts:287]: hmm, nvm, I think it is  good with gettter/setter for those classes

**Review by duanxianbo — APPROVED** [review]

File: src/libs/draw/page.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+export class Page {
+    private drawables: Drawable[] = [];
+
+    add(drawable: Drawable) {
+        this.drawables.push(drawable);
+    }
+
+    // remove(drawableId: number) {
+    //     this.drawables = this.drawables.filter(drawable => drawable.getId() !== drawableId);
+    // }
```
Comment (naturetime10) [inline src/libs/draw/page.ts]:
```suggestion
```

**Pull request merged**